org.eclipse.swt/Eclipse SWT/common/org/eclipse/swt/internal/DPIUtil.d

```diff
@@ -0,0 +1,178 @@
+import org.eclipse.swt.graphics.ImageData;
+import org.eclipse.swt.graphics.ImageDataProvider;
+import org.eclipse.swt.graphics.PaletteData;
+
+import java.lang.Math;
+import java.lang.System;
+import java.lang.String;
+import java.lang.util : Format;
+
+/**
+ * This class hold common constants and utility functions w.r.t. to SWT high DPI
+ * functionality.
+ * <p>
+ * The {@code autoScaleUp(..)} methods convert from API coordinates (in
+ * SWT points) to internal high DPI coordinates (in pixels) that interface with
+ * native widgets.
+ * </p>
+ * <p>
+ * The {@code autoScaleDown(..)} convert from high DPI pixels to API coordinates
+ * (in SWT points).
+ * </p>
+ *
+ * @since 3.105
+ */
+public class DPIUtil {
+package(org.eclipse.swt.internal):
+
+    private static const int DPI_ZOOM_100 = 96;
+
+    private static int deviceZoom = 100;
+    private static int nativeDeviceZoom = 100;
+
+    private static enum AutoScaleMethod { AUTO, NEAREST, SMOOTH }
+    private static AutoScaleMethod autoScaleMethodSetting = AutoScaleMethod.AUTO;
+    private static AutoScaleMethod autoScaleMethod = AutoScaleMethod.NEAREST;
+
+    /**
+     * System property that controls the autoScale functionality.
+     * <ul>
+     * <li><b>false</b>: deviceZoom is set to 100%</li>
+     * <li><b>integer</b>: deviceZoom depends on the current display resolution,
+     *     but only uses integer multiples of 100%. The detected native zoom is
+     *     generally rounded down (e.g. at 150%, will use 100%), unless close to
+     *     the next integer multiple (currently at 175%, will use 200%).</li>
+     * <li><b>integer200</b>: like <b>integer</b>, but the maximal zoom level is 200%.</li>
+     * <li><b>quarter</b>: deviceZoom depends on the current display resolution,
+     *     but only uses integer multiples of 25%. The detected native zoom is
+     *     rounded to the closest permissible value.</li>
+     * <li><b>exact</b>: deviceZoom uses the native zoom (with 1% as minimal
+     *     step).</li>
+     * <li><i>&lt;value&gt;</i>: deviceZoom uses the given integer value in
+     *     percent as zoom level.</li>
+     * </ul>
+     * The current default is "integer200".
+     */
+    private static const String SWT_AUTOSCALE = "swt.autoScale";
+
+    /**
+     * System property that controls the method for scaling images:
+     * <ul>
+     * <li>"nearest": nearest-neighbor interpolation, may look jagged</li>
+     * <li>"smooth": smooth edges, may look blurry</li>
+     * </ul>
+     * The current default is to use "nearest", except on
+     * GTK when the deviceZoom is not an integer multiple of 100%.
+     * The smooth strategy currently doesn't work on Win32 and Cocoa, see
+     * <a href="https://bugs.eclipse.org/493455">bug 493455</a>.
+     */
+    private static const String SWT_AUTOSCALE_METHOD = "swt.autoScale.method";
+    static this() {
+        String value = System.getProperty (SWT_AUTOSCALE_METHOD);
+        if (value !is null) {
```

Comment:
What if this the string has a zero length but is not null? Does it matter? Java and D has different behavior.

Reply:
At least for strings (since that's all I've tested), the behaviour is the same for both Java and D.

For example, in Java:

```java
public class Hello
{
   public static void main(String[] args)
     {
	String value = "";
	if (value != null)
	  System.out.println("value != null"); // This prints
	else
	  System.out.println("value == null");
	
	value = null;
	if (value != null)
	  System.out.println("value != null");
	else
	  System.out.println("value == null"); // This prints
     }   
}
```

Same behaviour in D:

```d
void main()
{
   import std.stdio : writeln;

   string value = "";
   if (value !is null)
     writeln("value !is null"); // This prints
   else
     writeln("value is null");
   
   value = null;
   if (value !is null)
     writeln("value !is null");
   else
     writeln("value is null"); // This prints
}

```